cot/src/common_types.rs

```diff
@@ -132,6 +132,233 @@ impl From<String> for Password {
         Self::new(password)
     }
 }
+/// A validated URL wrapper.
+///
+/// This structure ensures that the contained URL is correctly formatted and
+/// complies with standard URL syntax rules. It wraps [`url::Url`] to provide
+/// validation upon construction through methods like [`new`] and [`from_str`],
+/// and exposes useful methods for accessing URL components or converting the
+/// URL into different representations.
+///
+/// # Examples
+///
+/// ## Creating a Validated URL Using `new`
+///
+/// ```
+/// use cot::common_types::Url;
+///
+/// // Successful URL creation
+/// let url = Url::new("https://example.com").unwrap();
+///
+/// // Accessing the normalized URL string
+/// assert_eq!(url.as_str(), "https://example.com/");
+/// ```
+///
+/// ## Parsing a URL from a String Using `from_str`
+///
+/// ```
+/// use std::str::FromStr;
+///
+/// use cot::common_types::Url;
+///
+/// // Parse a valid URL string
+/// let url = Url::from_str("https://example.com").unwrap();
+///
+/// // Convert into owned string representation
+/// let url_string = url.into_string();
+/// assert_eq!(url_string, "https://example.com/");
+/// ```
+///
+/// # Behavior
+///
+/// - **Validation**: Both `new` and `from_str` ensure the input is a syntactically
+///   correct URL as defined by the WHATWG URL specification via the underlying
+///   [`url::Url`] parser.
+/// - **Normalization**: The internal URL is normalized (e.g., trailing slash added
+///   for HTTP URLs) during construction.
+#[derive(Clone, Debug)]
+pub struct Url(url::Url);
+
+impl Url {
+    /// Creates a new `Url` by parsing the input string.
+    ///
+    /// # Parameters
+    /// - `s`: The input string to parse as a URL
+    ///
+    /// # Returns
+    /// - `Ok(Url)`: Successfully parsed URL
+    /// - `Err(url::ParseError)`: If parsing fails
+    ///
+    /// # Errors
+    /// Returns `url::ParseError` if the input string is not a valid URL format.
+    ///
+    /// # Examples
+    /// ```
+    /// use cot::common_types::Url;
+    ///
+    /// let valid_url = Url::new("https://example.com").unwrap();
+    /// ```
+    pub fn new<S: AsRef<str>>(s: S) -> Result<Url, url::ParseError> {
+        url::Url::from_str(s.as_ref()).map(Self)
+    }
+
+    /// Returns a string slice reference to the URL's string representation.
+    ///
+    /// # Returns
+    /// - `&str`: A reference to the underlying URL string
```

Comment:
```suggestion
    /// Returns a string slice reference to the URL's string representation.
```

I think there's not much benefit in writing the same thing twice :wink:

And it will be more consistent, too, since we typically don't have a `# Returns` subsection in the API docs (unless there's a strong reason to include one).

This also applies to other methods.